cpp/src/cider/common/Arena.h

```diff
@@ -40,35 +54,137 @@ namespace cider::hashtable {
+      begin = reinterpret_cast<char*>(CiderDefaultAllocator::allocate(size_));
+      pos = begin;
+      end = begin + size_ - pad_right;
+      prev = prev_;
+    }
+
+    ~MemoryChunk() {
+      CiderDefaultAllocator::deallocate(reinterpret_cast<int8_t*>(begin), size());
+      delete prev;
+    }
+
+    size_t size() const { return end + pad_right - begin; }
+    size_t remaining() const { return end - pos; }
+  };
+
+  size_t growth_factor;
+  size_t linear_growth_threshold;
+
+  // Last contiguous MemoryChunk of memory.
+  MemoryChunk* head;
+  size_t size_in_bytes;
+  size_t page_size;
+
+  static size_t roundUpToPageSize(size_t s, size_t page_size) {
+    return (s + page_size - 1) / page_size * page_size;
+  }
+
+  // If MemoryChunks size is less than 'linear_growth_threshold', then use exponential
+  // growth, otherwise - linear growth
+  //  (to not allocate too much excessive memory).
+  size_t nextSize(size_t min_next_size) const {
+    size_t size_after_grow = 0;
+
+    if (head->size() < linear_growth_threshold) {
+      size_after_grow = std::max(min_next_size, head->size() * growth_factor);
+    } else {
+      // allocContinue() combined with linear growth results in quadratic
+      // behavior: we append the data by small amounts, and when it
+      // doesn't fit, we create a new MemoryChunk and copy all the previous data
+      // into it. The number of times we do this is directly proportional
+      // to the total size of data that is going to be serialized. To make
+      // the copying happen less often, round the next size up to the
+      // linear_growth_threshold.
+      size_after_grow =
+          ((min_next_size + linear_growth_threshold - 1) / linear_growth_threshold) *
+          linear_growth_threshold;
+    }
+
+    return roundUpToPageSize(size_after_grow, page_size);
+  }
 
-  /// If MemoryChunks size is less than 'linear_growth_threshold', then use exponential
-  /// growth, otherwise - linear growth
-  ///  (to not allocate too much excessive memory).
-  size_t nextSize(size_t min_next_size) const { return 0; }
+  // Add next contiguous MemoryChunk of memory with size not less than specified.
+  void NO_INLINE addMemoryChunk(size_t min_size) {
+    head = new MemoryChunk(nextSize(min_size + pad_right), head);
+    size_in_bytes += head->size();
+  }
 
-  /// Add next contiguous MemoryChunk of memory with size not less than specified.
-  void addMemoryChunk(size_t min_size) {}
+  friend class ArenaAllocator;
+  template <size_t>
+  friend class AlignedArenaAllocator;
 
  public:
-  /// Get piece of memory, without alignment.
-  char* alloc(size_t size) { return nullptr; }
+  explicit Arena(size_t initial_size_ = 4096,
+                 size_t growth_factor_ = 2,
+                 size_t linear_growth_threshold_ = 128 * 1024 * 1024)
+      : growth_factor(growth_factor_)
+      , linear_growth_threshold(linear_growth_threshold_)
+      , head(new MemoryChunk(initial_size_, nullptr))
```

Comment:
when will release head's resource?

Reply:
Memory will be reused if we don't need the objects anymore, so it won't be released until the task is completed.